src/burst_buffer/burst_buffer.lua

```diff
@@ -1,39 +1,722 @@
+--used by jobs.
+--
+--If pools are desired, they must be returned as the second return value
+--of this function. It must be a single JSON string representing the pools.
+--]]
+function slurm_bb_pools()
+
+	slurm.log_info("%s: slurm_bb_pools().", lua_script_name)
+
+	--This commented out code specifies pools in a file:
+	--local pools_file, pools
+	--pools_file = "/path/to/file"
+
+	--io.input(pools_file)
+	--pools = io.read("*all")
+	--slurm.log_info("Pools file:\n%s", pools)
+
+	--This specifies pools inline:
+	local pools
+	pools ="\
+{\
+\"pools\":\
+  [\
+    { \"id\":\"pool1\", \"quantity\":1000, \"granularity\":1024 },\
+    { \"id\":\"pool2\", \"quantity\":5, \"granularity\":2 },\
+    { \"id\":\"pool3\", \"quantity\":4, \"granularity\":1 },\
+    { \"id\":\"pool4\", \"quantity\":25000, \"granularity\":1 }\
+  ]\
+}"
+
+	return slurm.SUCCESS, pools
+end
+
+--[[
+--slurm_bb_job_teardown
+--
+--This function is called asynchronously and is not required to return quickly.
+--This function is normally called after the job completes (or is cancelled).
+--]]
+function slurm_bb_job_teardown(job_id, job_script, hurry)
+	slurm.log_info("%s: slurm_bb_job_teardown(). job id:%s, job script:%s, hurry:%s",
+		lua_script_name, job_id, job_script, hurry)
+
+	local hurry_flag = false
+	if hurry == "true" then
+		hurry_flag = true
+	end
+	local workflow = DWS(make_workflow_name(job_id))
+	local done, err = workflow:set_workflow_state_and_wait("Teardown", hurry_flag)
+	if done == false then
+		slurm.log_error("%s: slurm_bb_job_teardown(). %s", lua_script_name, err)
+		return slurm.ERROR, err
+	end
+
+	done, err = workflow:delete()
+	if done == false then
+		slurm.log_error("%s: slurm_bb_job_teardown(). %s", lua_script_name, err)
+		return slurm.ERROR, err
+	end
+
+	return slurm.SUCCESS
+end
+
+--[[
+--slurm_bb_setup
+--
+--This function is called asynchronously and is not required to return quickly.
+--This function is called while the job is pending.
+--]]
+function slurm_bb_setup(job_id, uid, gid, pool, bb_size, job_script)
+	slurm.log_info("%s: slurm_bb_setup(). job id:%s, uid: %s, gid:%s, pool:%s, size:%s, job script:%s",
+		lua_script_name, job_id, uid, gid, pool, bb_size, job_script)
+
+	-- See the notes in slurm_bb_process() for an explanation about why we
+	-- create the Workflow resource here rather than look up an existing
+	-- resource.
+
+	local workflow_name = make_workflow_name(job_id)
+	local workflow = DWS(workflow_name)
+	local done, err = make_workflow(workflow, job_script, job_id, uid, gid)
+	if done == false then
+		slurm.log_error("%s: slurm_bb_setup(). %s", lua_script_name, err)
+		return slurm.ERROR, err
+	end
+
+	-- Wait for proposal state to complete, or pick up any error that may
+	-- be waiting in the Workflow.
+	done, err = workflow:wait_for_status_complete(60)
+	if done == false then
+		slurm.log_error("%s: slurm_bb_setup(). %s", lua_script_name, err)
```

Comment:
We need more information here about what we are waiting for. An error here would just say "exceeded max wait time" and would not indicate if we timed out during the creation or setup step.